### PR TITLE
feat: Added support for getting container ids from ECS metadata API

### LINKED ISSUE #2276: Add support for fetching container identifier in ECS Fargate
When an application is running inside of a Docker container, on a Linux guest, our agent is able to retrieve an identifier for that container via cgroups that are populated by Docker. However, when the host is AWS ECS Fargate, those cgroups are not present within the guest. Instead, a [metadata API](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ec2-metadata.html) that is available to the guest should be consulted.

We need to:

1. Update the following function to utilize this metadata endpoint if the cgroups are not available: https://github.com/newrelic/node-newrelic/blob/07a841b9acf0a0dcb1a495c5cb0ee8f77e283fc9/lib/utilization/docker-info.js#L62-L88
2. We should first attempt to look for the API endpoint through the URI available in the `ECS_CONTAINER_METADATA_URI_V4` environment variable. If that environment variable does not exist, try the `ECS_CONTAINER_METADATA_URI` environment variable. If neither exist, exit with a failure as the function currently does.

+ Internal spec: https://source.datanerd.us/agents/agent-specs/blob/12fa39c941badb0de502b07270ab061fd3417cf5/Utilization.md#docker-with-aws-ecs-fargate
+ .Net agent implementation: https://github.com/newrelic/newrelic-dotnet-agent/pull/2481/files

### PR DESCRIPTION
**Pull request by jsumners-nr**
This PR resolves #2276.

### CODE REVIEW

File: lib/utilization/docker-info.js [hunk 1]
```diff
@@ -5,61 +5,162 @@
...
+ *
+ * @see https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ec2-metadata.html
+ *
+ * @returns {boolean}
+ */
+function hasAwsContainerApi() {
+  if (process.env.ECS_CONTAINER_METADATA_URI_V4 != null) {
+    return true
   }
+  return process.env.ECS_CONTAINER_METADATA_URI != null
```
Comment (bizob2828) [inline lib/utilization/docker-info.js:109]:
```suggestion
  return process.env.ECS_CONTAINER_METADATA_URI_V4 != null || process.env.ECS_CONTAINER_METADATA_URI != null
```

File: lib/utilization/docker-info.js [hunk 2]
```diff
@@ -5,61 +5,162 @@
...
 const CGROUPS_V2_PATH = '/proc/self/mountinfo'
+const BOOT_ID_PROC_FILE = '/proc/sys/kernel/random/boot_id'
 
 module.exports.getVendorInfo = fetchDockerVendorInfo
 module.exports.clearVendorCache = function clearDockerVendorCache() {
   vendorInfo = null
 }
 
-module.exports.getBootId = function getBootId(agent, callback) {
+module.exports.getBootId = function getBootId(agent, callback, logger = log) {
```
Comment (bizob2828) [inline lib/utilization/docker-info.js:25]: it's fine but seems like a lot of signature updates to inject a logger.  if you look through our unit tests we use proxyquire to inject mock loggers

File: test/unit/utilization/docker-info.test.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+const helper = require('../../lib/agent_helper')
+const standardResponse = require('./aws-ecs-api-response.json')
+const { getBootId } = require('../../../lib/utilization/docker-info')
+
+tap.beforeEach(async (t) => {
+  t.context.orig = {
+    fs_access: fs.access,
+    os_platform: os.platform
+  }
+  fs.access = (file, mode, cb) => {
```
Comment (bizob2828) [inline test/unit/utilization/docker-info.test.js:21]: this is purely informational but a lot of our tests use sinon for mocking.

`sinon.stub(fs, 'access').yields(Error('no proc file'))`

**Review by bizob2828 — APPROVED** [review]
tiny little suggestion on the checking for the ECS env vars but up to you if you want to address it

File: lib/utilization/docker-info.js (hunk 1 above)
Comment (jsumners-nr) [reply lib/utilization/docker-info.js:109]: For me, it's clearer as it is written that we are preferring the `V4` version and then falling back to the other.

File: lib/utilization/docker-info.js (hunk 2 above)
Comment (jsumners-nr) [reply lib/utilization/docker-info.js:25]: I think being able to inject the parameter is a lot cleaner than using proxyquire. We don't need to couple the implementation of the whole module to the tests (the tests needing to know how the logger is imported and used in the code).

**Pull request merged**